airflow/contrib/executors/kubernetes_executor.py

```diff
@@ -651,14 +651,27 @@ def _labels_to_key(self, labels):
                     'severe performance regressions. Please see '
                     '<https://kubernetes.io/docs/concepts/overview/working-with-objects'
                     '/labels/#syntax-and-character-set>. '
-                    'Given dag_id: %s, task_id: %s', task_id, dag_id
+                    'Given dag_id: %s, task_id: %s', dag_id, task_id
+                )
+            # attempt to avoid the extremely slow path by seeing if dag_id is valid to be added to
+            # search query
+            # if the dag_id comes back from the `safe_label` code unchanged, it is already a valid label
+            if self._make_safe_label_value(dag_id) == dag_id:
+                self.log.info('Attempting enhanced slow-path lookup as dag_id appears to be a valid K8 label')
+                tasks = (
+                    session
+                    .query(TaskInstance)
+                    .filter_by(dag_id=dag_id, execution_date=ex_time)
+                    .all()
+                )
+            else:
+                self.log.warning('Normal slow-path lookup must be used to find tasks')
+                tasks = (
+                    session
+                    .query(TaskInstance)
+                    .filter_by(execution_date=ex_time)
+                    .all()
```

Comment:
I haven't checked how the metadata looks like internally but is there anything we could add (filters) to this query to make it more performant?

Reply:
That's essentially what my PR is doing, adding in the dag_id when it would match.  

The only thing it uses to match pods to TaskInstances are the labels for dag_id and task_id from the pod and the execution date.   The trouble occurs when it can't fit either the dag_id or task_id into the K8 pod label (usually due to excessive string length > 63); instead it sets the label to be a hashed version.  When the pod completes it needs to do a lookup from labels -> TaskInstance, but the TaskInstance doesn't have the hashed values, it has the actual values.  So all it goes on is the execution date to get ALL tasks with the same date and then iterates, hashes and compares each. 

The flaw is that the algorithm assumes if ANY label is hashed, then it needs to revert to that worst case.  My limited observation suggests that it is usually ONLY the task_id that is too long so this PR adds another case to add the dag_id as a filter which speeds things up dramatically.  That way we only need to retrieve all tasks in a specific DAG for a given execution period which is much smaller and faster than ALL DAGs for a given execution period.

I'm not sure if there is any additional metadata from the pods that could be used as a filter to map back to TaskInstances in the worst case of both dag_id and task_id being hashed, I'll have to look into that a bit.

I thought about maybe trying to add an index to speed things up, but I'm not very familiar with how to do so in SQLAlchemy so I didn't get very far with that yet.

Reply:
> That's essentially what my PR is doing, adding in the dag_id when it would match.

Yes, but I mean for the worst case, to make it less bad if possible. I guess no, but I was wondering if an extra filter could help us prune some branches.

Reply:
Worth a look after we get this deployed, maybe we can figure something out